src/collapse/perf.rs

```diff
@@ -412,3 +560,129 @@ fn tidy_java(mut func: String) -> String {
 
     func
 }
+
+// Removes discriminator markers.
+// Equivalent to s/ \(discriminator \S+\)//
+fn remove_discriminator(s: Cow<str>) -> Cow<str> {
+    let disc = " (discriminator ";
+    if let Some(start) = s.find(disc) {
+        let rest = &s[start + disc.len()..];
+        if let Some(end) = rest.find(')') {
+            if rest[..end].chars().all(|c| !c.is_whitespace()) {
+                return Cow::from(format!(
+                    "{}{}",
+                    &s[0..start],
+                    &s[start + disc.len() + end + 1..s.len()]
+                ));
+            }
+        }
+    }
+    s
+}
+
+// Removes Rust hash from function name if it exists.
+// When demangling Rust names, addr2line in binutils removes the hash,
+// but the Rust addr2line crate does not. Calling this function
+// after demangling will make it match binutils.
+fn strip_rust_hash(func: &str) -> &str {
+    if let Some(idx) = func.rfind("::") {
+        if is_rust_hash(&func[idx + 2..]) {
+            return &func[0..idx];
+        }
+    }
+    func
+}
+
+// Rust hashes are hex digits with an `h` prepended.
+fn is_rust_hash(s: &str) -> bool {
+    s.starts_with('h') && s[1..].chars().all(|c| c.is_digit(16))
+}
+
+// Returns whether the module can be un-inlined.
+//
+// Equivalent to the following perl:
+//     $mod !~ m/(perf-\d+.map|kernel\.|\[[^\]]+\])/;
+fn can_un_inline(module: &str) -> bool {
+    // /perf-\d+.map/
+    if let Some(perf) = module.find("perf-") {
+        let rest = &module[perf + 5..];
+        if let Some(map) = rest.find(".map") {
+            if rest[..map].chars().all(|c| c.is_ascii_digit()) {
+                return false;
+            }
+        }
+    }
+
+    // /kernel\./
+    if module.find("kernel.").is_some() {
+        return false;
+    }
+
+    // /\[[^\]]+\]/
+    let mut start = 0;
+    while start < module.len() {
+        if let Some(open) = module[start..].find('[') {
+            let rest = &module[open + 1..];
+            if let Some(close) = rest.find(']') {
+                if close > 1 && rest[..close].chars().all(|c| c != ']') {
```

Comment:
Won't this necessarily be true since `.find` will find the _next_ occurrence of `]`?